### PR TITLE
DAOS-15269 cq: merge GHA lint checks into linting.yml

### PR DESCRIPTION
**Pull request by daltonbohning**
Skip-test: true
Skip-tests: true
Skip-unit-tests: true
Skip-fault-injection-test: true

Merge these lint checks into linting.yml
- doxygen.yml
- flake.yml
- pylint.yml

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by ashleypittman — APPROVED** [review]
Awesome.

Once https://github.com/daos-stack/daos/pull/13833 is working we could probably make the doxygen one a unit test.

**Comment by daltonbohning** [comment]
Also merged the pylint check

**Comment by daltonbohning** [comment]
@ashleypittman Should we also merge codespell and clang-format?

**Comment by daltonbohning** [comment]
> Here is the [documentation](https://github.com/ossf/scorecard/blob/54690b41ae065b634a2d002007e3f95547c2deab/docs/checks.md#token-permissions) that IPAS referred to to address these/this.

And the GHA documentation for setting permissions
https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs

File: .github/workflows/linting.yml [hunk 1]
```diff
@@ -1,5 +1,7 @@
 name: Linting
 
+permissions: read-all
```
Comment (daltonbohning) [inline .github/workflows/linting.yml:3]: There are a lot of fine-grained permissions here, which if set would technically be more secure:
https://docs.github.com/en/actions/using-jobs/assigning-permissions-to-jobs

Comment (brianjmurrell) [reply .github/workflows/linting.yml:3]: Yes, indeed.  I suppose you are referring to the ability to set some of the finer grained permissions to `none` as I think effectively what you have done here is restricted them all to `read`, which is still quite a bit better than the default: https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token

Comment (daltonbohning) [reply .github/workflows/linting.yml:3]: Right. Copy-pasted from the link I posted

> You can define the access that the GITHUB_TOKEN will permit by specifying read, write, or none as the value of the available scopes within the permissions key.

```
permissions:
  actions: read|write|none
  checks: read|write|none
  contents: read|write|none
  deployments: read|write|none
  id-token: read|write|none
  issues: read|write|none
  discussions: read|write|none
  packages: read|write|none
  pages: read|write|none
  pull-requests: read|write|none
  repository-projects: read|write|none
  security-events: read|write|none
  statuses: read|write|none
```
> If you specify the access for any of these scopes, all of those that are not specified are set to none.

Comment (daltonbohning) [reply .github/workflows/linting.yml:3]: So, for example, we should be able to do:
```
permissions:
  actions: read
```
And everything else will be `none`.

But that requires knowing exactly which of those permissions the actions need

Comment (brianjmurrell) [reply .github/workflows/linting.yml:3]: Exactly!

Comment (daltonbohning) [reply .github/workflows/linting.yml:3]: Or maybe we should be setting EACH read permission at the top level as-needed?

Comment (daltonbohning) [reply .github/workflows/linting.yml:3]: > Remediation steps

> Set top-level permissions as read-all or contents: read as described in GitHub's [documentation](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#permissions).
Set any required write permissions at the job-level. Only set the permissions required for that job; do not set permissions: write-all at the job level.
To help determine the permissions needed for your workflows, you may use [StepSecurity's online tool](https://app.stepsecurity.io/secureworkflow/) by ticking the "Restrict permissions for GITHUB_TOKEN". You may also tick the "Pin actions to a full length commit SHA" to fix issues found by the Pinned-dependencies check

Comment (daltonbohning) [reply .github/workflows/linting.yml:3]: So I think what I did here is sufficient

**Review by brianjmurrell — APPROVED** [review]

**Comment by daltonbohning** [comment]
> > If we do not fix these by 2024-04-01 IPAS are going to mark our project as unmaintained and then "archive" the repo, setting it to read-only.
> 
> That sounds like a really big deal that we need to address immediately. Do we have proper urgency for this at the team level?

@brianjmurrell Do we have tracking for this? I'm concerned because at least all of our other GHA are not setting permissions. And AFAIK, there is no visible work to update them

**Comment by daltonbohning** [comment]
@brianjmurrell I don't think the permissions are as simple as just `read-all` because we need some write permission for actions to annotate the PR on failure. But that needs more testing and understanding.
Also, running this workflow through the [IPAS tool](https://app.stepsecurity.io/secureworkflow) flags a bunch of other issues, so in general I think these security updates need to be tracked separately

**Review by ashleypittman — DISMISSED** [review]

**Comment by brianjmurrell** [comment]
> @brianjmurrell Do we have tracking for this? I'm concerned because at least all of our other GHA are not setting permissions. And AFAIK, there is no visible work to update them

Lots of folks are on the notification list for these warnings all the way up to Brad so visibility is not a problem.

> @brianjmurrell I don't think the permissions are as simple as just `read-all`

I'm not at all surprised.

> because we need some write permission for actions to annotate the PR on failure.

Yup.  At least.

> But that needs more testing and understanding. Also, running this workflow through the [IPAS tool](https://app.stepsecurity.io/secureworkflow) flags a bunch of other issues,

With various enforcement timelines, yes.  The one you are hopefully addressing is the soonest.

> so in general I think these security updates need to be tracked separately

As I said above, there is no lack of visibility.  What there seems to be a lack of is ownership of the many workflows involved by those that know what they are doing and what is expected of them who can set the permissions accordingly.

**Comment by daltonbohning** [comment]
@brianjmurrell FYI, check out the OSSF scorecard in action! I can address all of these on this PR if you don't already have one?

**Comment by daltonbohning** [comment]
I fixed the simple SHA pins but the `pipCommand not pinned by hash` I think want us to pin every python version in the requirements file. That needs more thought

**Comment by brianjmurrell** [comment]
> @brianjmurrell FYI, check out the OSSF scorecard in action! I can address all of these on this PR if you don't already have one?

We are going to punt on that one for now as we aren't going to be held accountable for it for a few quarters still and I'm frankly suspicious of the feasibility of it.

> I fixed the simple SHA pins but the `pipCommand not pinned by hash` I think want us to pin every python version in the requirements file. That needs more thought

Indeed!  This is an example of my suspicions.  Right now the security dashboard is telling us that we have ~111 unpinned dependencies.  The idea that we have to track all 111 (and more as the list grows) of these and move them along manually, in order to, for example, pick up bug (and perhaps even security) fixes seems somewhat unworkable.

So, TL;DR: I would not worry about the pinning at this point in time.

**Comment by daltonbohning** [comment]
> > @brianjmurrell FYI, check out the OSSF scorecard in action! I can address all of these on this PR if you don't already have one?
> 
> We are going to punt on that one for now as we aren't going to be held accountable for it for a few quarters still and I'm frankly suspicious of the feasibility of it.
> 
> > I fixed the simple SHA pins but the `pipCommand not pinned by hash` I think want us to pin every python version in the requirements file. That needs more thought
> 
> Indeed! This is an example of my suspicions. Right now the security dashboard is telling us that we have ~111 unpinned dependencies. The idea that we have to track all 111 (and more as the list grows) of these and move them along manually, in order to, for example, pick up bug (and perhaps even security) fixes seems somewhat unworkable.
> 
> So, TL;DR: I would not worry about the pinning at this point in time. They are _medium_ and I don't think anything less than _high_ marks the PR with -1.

Yeah, it seems they want us to use some tool (I think provided?) to check and update dependencies. Pinning the SHA for workflows isn't so bad because we were already pinning the version. So unless there's an issue here, I'll wrap up this PR where it is.

**Review by brianjmurrell — DISMISSED** [review]

File: .github/workflows/linting.yml [hunk 2]
```diff
@@ -68,6 +68,97 @@ jobs:
...
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11  # v4.1.1
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
+      - uses: actions/setup-python@82c7e631bb3cdc910f68e0081d67478d79c6982d  # v5.1.0
+        with:
+          python-version: '3.11'
+      - name: Install core python packages
+        run: python3 -m pip install --requirement requirements.txt
+      - name: Install extra python packages
+        run: python3 -m pip install --requirement utils/cq/requirements.txt
```
Comment (ashleypittman) [inline .github/workflows/linting.yml:143]: Not a regression but `requirements.txt` pulls in `utils/cq/requirements.txt` now anyway so this run step isn't needed and could be removed.

Comment (daltonbohning) [reply .github/workflows/linting.yml]: Done

File: .github/workflows/linting.yml [hunk 3]
```diff
@@ -71,3 +71,94 @@
...
+        with:
+          ref: ${{ github.event.pull_request.head.sha }}
+      - uses: actions/setup-python@v5
+        with:
+          python-version: '3.11'
+      - name: Install core python packages
+        run: python3 -m pip install --requirement requirements.txt
+      - name: Install extra python packages
+        run: python3 -m pip install --requirement utils/cq/requirements.txt
+      - name: Install enchant
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply .github/workflows/linting.yml:142]: I think as an experiment we should dismiss these ones as "Used in tests" as this code only runs on github hosted runners.

**Review by ashleypittman — APPROVED** [review]

File: .github/workflows/linting.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply .github/workflows/linting.yml:142]: That will only dismiss the GitHub alert raised.  It will do nothing about the Scorecard report and IPAS' general requirement that **every single** dependency be pinned to a git hash.

**Review by brianjmurrell — APPROVED** [review]

File: .github/workflows/linting.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (daltonbohning) [reply .github/workflows/linting.yml:142]: But then shouldn't we dismiss all warnings for all GHA that run on GH runners? Then what's the point  in the OSSF requirements?
Not questioning your suggestion, but rather the feasibility of all of this as @brianjmurrell pointed out.

This dismissal description doesn't make sense to me
> These alerts are not in production code

None of our GHA are used in production code. I'm not sure what a "production GHA" would even be

[the remark this answers is not in the document]
Comment (ashleypittman) [reply .github/workflows/linting.yml:142]: I was building on what I thought you'd said about these being warnings, not errors.

> That will only dismiss the GitHub alert raised. It will do nothing about the Scorecard report and IPAS' general requirement that **every single** dependency be pinned to a git hash.

I was building on what I thought you'd said about these being warnings, not errors.

Let's leave this for now then pending further investigation.

**Comment by daltonbohning** [comment]
CI testing skipped since only GHA lint checks were modified.
